Everycodeacademy

xenex90 opened the pull request:
@f-lab-troy 멘토님

webComplier 초기 서버상에서만 작동하는 프로토 타입 제작 하였습니다.
아래 코드로 테스트 하였을 때

java 파일 생성 - 해당 java 파일 컴파일 하여 class 파일로 생성 - 생성된 class 파일 읽어서 동작 - 이 과정에서 생성 된 파일 권한 변경 후 삭제
하는 순으로 동작합니다.

현재는 javaBodyString 파라미터명으로 값을 보내면 해당 값이 콘솔에 뜨도록 코딩되어 있습니다.
개발 중인데 중간 중간 PR 받고 싶어서 PUSH 하였습니다.

추후 좀 더 계속 손 볼 예정입니다.

계획 중인 것은 파일 명을 랜덤이나 해쉬 값으로 생성해서 여러 API요청이 와도 파일 명이 중복되지 않도록 할 예정이고
API로 값이나 데이터를 보낼 수 있도록 할 수 있는 UI 추가 예정,
컴파일 과정에서 에러나 문제점에 관해서 전혀 고려되어 있지 않은 상태여서 이 부분도 생각해야합니다.


`{
    "javaBodyString": "test sucess",
    "username": "youngju",
    "parameterString": "args"
}`

![image](https://github.com/f-lab-edu/everybodycodeacademy/assets/56420135/6ed81195-6798-4db4-9cb5-80f76f5656f3)


---
---

멘토님 추가 commit & push 하여서
동시성 이슈 (Race Conditon) 관련 코드 추가하였습니다.
Sycrhonized 키워드를 써서 해결해 보았습니다.
멀티스레드 일떄는 이 방법이 통하지만
여러 자바 애플리케이션에서 DB로 접근해서 데이터 처리할 떄는
트랜잭션 등 다른 방법을 써야 한다고 해서 다른 방법도 해보겠습니다.

xenex90 committed:
Initical codeacademy commit

첫 프로젝트 커밋입니다.

xenex90 committed:
Add board list, write, read, modify

xenex90 committed:
Add SpringDoc to view swagger_ui

-SpringDoc 의존성 추가
-Resquest regDate,modDate 변수 Stirng타입 -> LocalDateTime타입으로 형변

xenex90 committed:
Remove unusing controller_BoardController

-UI적으로 동작 필요했던 Controller가 필요없다고 판단하여 삭제

xenex90 committed:
Upload bootstrap ui original zip_file

-현재 필수는 아니지만 UI구현에 필요한 bootstrap 기반 UI 구현에 필요한 원본 zip파일 업로드 (license 획득)

xenex90 committed:
Move and remove ui_files

-동작에 필요한 몇몇 파일들 templates -> static
-파일 수가 너무 많아 우선 파일들 대거 삭제 (동작엔 영향은 크게 없지만 이미지 등 대부분 삭제) (필요시 추후 추가)

xenex90 committed:
Move and remove ui_files_2

-실수로 Move and remove ui_files commit에 필요한 모든 파일을 다 선택/반영하지 않아 편의상 추가 commit으로 진행합니다.
-동작에 필요한 몇몇 파일들 templates -> static
-파일 수가 너무 많아 우선 파일들 대거 삭제 (동작엔 영향은 크게 없지만 이미지 등 대부분 삭제) (필요시 추후 추가)

xenex90 committed:
Remove all ui files

-UI 관련 html, img, js, zip 등 삭제 하였습니다.

xenex90 committed:
Add docker setting file, docker_compose setting file and modify build.gradle

-docker image 생성 설정 파일 추가 - Dockerfile
-docker compose 설정 파일 추가 - docker-compose.yml
-jar 파일 생성 할 경우를 위한 main-class 위치 지정 - build.gradle

도커와 도커 컴포즈로 구동 위한 설정파일을 추가, 이를 위해 build.gradle 수정

xenex90 committed:
Merge branch 'main' into everycodeacademy

xenex90 committed:
Modify build.gradle to build jar file

- gradle build command: war -> executable jar

xenex90 committed:
remove Silicon-2.zip files again

xenex90 committed:
test ci 1

xenex90 committed:
test ci 1

xenex90 committed:
test ci 3

xenex90 committed:
test ci 3 (#16)

xenex90 committed:
test ci 4

xenex90 committed:
test ci 4

xenex90 committed:
test ci 5

xenex90 committed:
test ci 5

xenex90 committed:
Merge branch 'everycodeacademy_cicd' into everycodeacademy

xenex90 committed:
modify application.properties and Dockerfile for EC2 CICD(github action)

- application.properties - Mysql server ip 반영
- Dockerfile - scouter agent(java) 반영

xenex90 committed:
Merge remote-tracking branch 'origin/everycodeacademy' into everycodeacademy

xenex90 committed:
Update application.properties

xenex90 committed:
Update application.properties

xenex90 committed:
Merge branch 'main' into everycodeacademy

xenex90 committed:
Add load Test Method

Transactional 어노테이션 추가 및
Isolation : READ_UNCOMMITTED, READ_UNCOMMITTED, REPEATABLE_READ, SERIALIZABLE 4가지 태그 test용으로 작성

xenex90 committed:
Add load Test Method

Transactional 어노테이션 추가 및
Isolation : READ_UNCOMMITTED, READ_UNCOMMITTED, REPEATABLE_READ, SERIALIZABLE 4가지 태그 test용으로 작성

xenex90 committed:
move load Test Method & class

- 로드 테스트 관련 메소드들 따로 class 생성 후 이동
- 기존 소스코드 정렬

xenex90 committed:
Merge branch 'main' into everycodeacademy

xenex90 committed:
Add Simple webCompiler test version

로컬환경에서 테스트 된
java -> class -> String 출력 -> 파일 삭제로
진행되는 CompileAPI 클래스 추가 및
관련 Service 클래스 추가하였습니다.

xenex90 committed:
Add Simple webCompiler test version

Simple webCompiler 개발하면서 제작한 테스트 케이스 들입니다.

xenex90 commented on everycodeacademy/src/test/java/com/every/everycodeacademy/API/compileAPITest.java:
```diff
@@ -0,0 +1,51 @@
+import org.springframework.boot.test.context.SpringBootTest;
+
+import static org.junit.jupiter.api.Assertions.*;
+
```
테스트 케이스 입니다.

xenex90 commented on everycodeacademy/src/test/java/com/every/everycodeacademy/compile/ChangeClassFilePermissions.java:
```diff
@@ -0,0 +1,33 @@
+import java.util.Set;
+import org.junit.jupiter.api.Test;
+import org.springframework.boot.test.context.SpringBootTest;
+
```
테스트 케이스 입니다.

xenex90 commented on everycodeacademy/src/test/java/com/every/everycodeacademy/compile/DeleteClassFile.java:
```diff
@@ -0,0 +1,35 @@
+import org.mockito.InjectMocks;
+import org.mockito.MockitoAnnotations;
+import org.springframework.boot.test.context.SpringBootTest;
+
```
테스트 케이스 입니다.

xenex90 commented on everycodeacademy/src/test/java/com/every/everycodeacademy/compile/runCompileTest.java:
```diff
@@ -0,0 +1,59 @@
+import org.mockito.MockitoAnnotations;
+import org.springframework.boot.test.context.SpringBootTest;
+import static org.junit.jupiter.api.Assertions.*;
+
```
테스트 케이스 입니다.

xenex90 commented on everycodeacademy/src/test/java/com/every/everycodeacademy/compile/runCompileTest.java:
```diff
@@ -0,0 +1,59 @@
+  void setUp() {
+    MockitoAnnotations.openMocks(this);
+  }
+
```
테스트 케이스 입니다.

xenex90 commented on everycodeacademy/src/main/java/com/every/everycodeacademy/compile/JavaCompileService.java:
```diff
@@ -0,0 +1,107 @@
+import java.nio.file.attribute.PosixFilePermissions;
+import java.util.Set;
+import org.springframework.stereotype.Service;
+
```
파일 권한 변경 및 파일 삭제 메소드는 JavaCompileService에 추가하였습니다.

xenex90 commented on everycodeacademy/src/main/java/com/every/everycodeacademy/board/Board.java:
```diff
@@ -20,34 +20,23 @@
 @ToString
 public class Board {
 
```
이거 왜 이런지는 모르겠는데 코드가 전 변경 안한 것 같은데 변경되네요....

f-lab-troy replied:
포멧팅을 다시 한 경우 whitespace때문에 발생하곤 합니다

xenex90 replied:
감사합니다

xenex90 commented on everycodeacademy/src/main/java/com/every/everycodeacademy/API/CompileAPI.java:
```diff
@@ -0,0 +1,90 @@
+public class CompileAPI {
+
+  @Autowired JavaCompileService javaCompileService;
+
```
getJavaStringToCompile 메소드만 봐주시면 감사하겠습니다
전 그냥 기능 구현에만 신경썼는데
제가 놓친 부분이 있을까요??

xenex90 commented on everycodeacademy/src/main/java/com/every/everycodeacademy/compile/JavaCompile.java:
```diff
@@ -0,0 +1,18 @@
+@Setter
+public class JavaCompile {
+    LocalDateTime now = LocalDateTime.now();
+
```
API에서 쓰는 값을 매핑하는 JavaCompile 클래스 입니다.
VO? DTO 그런 기능을 합니다.

xenex90 commented on everycodeacademy/src/main/java/com/every/everycodeacademy/test/LoadTestAPI.java:
```diff
@@ -0,0 +1,65 @@
+public class LoadTestAPI {
+
+  private final LoadTestService loadTestService;
+
```
저번에 추가한 부하 관련 테스트 할려고 추가한 메소드들입니다.
제대로 동작 할련지는 모르겠습니다.
우선적으로 일단 추가해 보았습니다.

xenex90 commented on everycodeacademy/src/main/java/com/every/everycodeacademy/test/LoadTestService.java:
```diff
@@ -0,0 +1,80 @@
+    boardRepository.save(board);
+    System.out.println("2번 업데이트 됨");
+  }
+
```
트랜잭션에 옵션을 여러가지로 달아보았습니다.
마찬가지로 제대로 동작할지는 모르고 구상만 해보고 그걸 코딩하였습니다.
맞는 방법인지 더 알아보고 필요시 수정하겠습니다.

xenex90 committed:
Add concurrency race condition Test case

- 동시성 이슈 (race condition) 발생하는 코드에서
(JPA) 값 Read - +1 Update - Write 과정의 메소드에
synchronized 키워드 추가한 테스크 케이스 및 관련 코드

xenex90 commented on everycodeacademy/src/test/java/com/every/everycodeacademy/concurrency/MakeRunRaceTest.java:
```diff
@@ -0,0 +1,64 @@
+public class MakeRunRaceTest {
+  @Autowired private ConcurrencyRepository concurrencyRepository;
+  @Autowired private ConcurrencyService concurrencyService;
+
```
추가한 Race Condtion 발생 Test Case 입니다.

여러 쓰레드를 동시에 돌려서 동시성 이슈를 발생하도록 되어 있습니다. 현재 200개 돌리도록 되어 있어서
DB에 200이 찍어야 되지만 이슈가 발생하면 그 이하의 숫자가 발생합니다.

현재 다른 class의 메소드에 synchronized 키워드 추가하여 동시성 이슈가 발생하지 않도록 되어 있습니다.
더 큰 숫자로 테스트 하면 어떨지는 모르겠으나 200개까지 몇번의 테스트에서는 동시성 이슈(Race Condtion)이 발생하지는 않아
DB의 값이 200으로 찍힙니다.

xenex90 commented on everycodeacademy/src/main/java/com/every/everycodeacademy/concurrency/ConcurrencyServiceImpl.java:
```diff
@@ -0,0 +1,31 @@
+  //private final Object lock = new Object(); // 동기화를 위한 락 객체
+
+  @Autowired private ConcurrencyRepository concurrencyRepository;
+
```
동시성 이슈 해결을 위한 방법으로 synchronized를 추가하여 해결하였습니다.

xenex90 commented on everycodeacademy/src/main/java/com/every/everycodeacademy/concurrency/Concurrency.java:
```diff
@@ -0,0 +1,15 @@
+package com.every.everycodeacademy.concurrency;
+
```
테스트를 위한 엔티티 입니다.

xenex90 commented on everycodeacademy/src/main/java/com/every/everycodeacademy/concurrency/ConcurrencyRepository.java:
```diff
@@ -0,0 +1,7 @@
+
+import org.springframework.data.jpa.repository.JpaRepository;
+import org.springframework.stereotype.Repository;
+
```
JPA로 하여보았습니다.

f-lab-troy commented on everycodeacademy/src/main/java/com/every/everycodeacademy/API/BoardAPI.java:
```diff
@@ -5,27 +5,28 @@
-    @ApiResponse(responseCode = "200", description = "success"),
-    @ApiResponse(responseCode = "400", description = "bad request"),
-    @ApiResponse(responseCode = "404", description = "not found"),
-    @ApiResponse(responseCode = "500", description = "internal sever error")
```
Http default status 코드와 에러메세지로 보이는데 따로 명시해주신 이유가 있을까요?

xenex90 replied:
API문서에서 볼 때 코드에서 명시해 주면 좋다고 해서 적었는데
200, 400 같은 코드 같이 기본적으로 description도 명시해 주는 것이 있어서
지금으로서는 제가 적은 description은 딱히 기능적이론 더 유용하지는 않은 것 같습니다.
일단 지워도 될 것 같아서 지우고 나중에 필요하면 다시 적거나 수정하겠습니다.

f-lab-troy commented on everycodeacademy/src/main/java/com/every/everycodeacademy/API/CompileAPI.java:
```diff
@@ -0,0 +1,90 @@
+
+    JavaCompiler webCompiler = ToolProvider.getSystemJavaCompiler();
+
+    String filePath = "webCompile.java";
```
이처럼 정해진 값은 상수로 할당하는 방식을 추천드립니다

xenex90 replied:
네 아래에서 말씀하신 바와 같이
상수나 설정파일로 관리하겠습니다.

f-lab-troy commented on everycodeacademy/src/main/java/com/every/everycodeacademy/API/CompileAPI.java:
```diff
@@ -0,0 +1,90 @@
+
+    // javaCompile.setJavaBodyString("hihi1231231231");
+    // javaCompile.setParameterString("args");
+    javaCompile.setJavaFullCompile(
```
자바 코드 한줄만 컴파일 하는 의도로 보입니다. 
보통 온라인컴파일러는 class전체를 컴파일 하도록 하는 것으로 알고 있어 고려해보셔도 좋을 것 같습니다.

xenex90 replied:
네 알겠습니다
참고하겠습니다

f-lab-troy commented on everycodeacademy/src/main/java/com/every/everycodeacademy/API/CompileAPI.java:
```diff
@@ -0,0 +1,90 @@
+
+    try (BufferedWriter writer = new BufferedWriter(new FileWriter(filePath))) {
+      writer.write(javaCompile.getJavaFullCompile());
+      System.out.println("소스 코드가 " + filePath + "에 저장되었습니다.");
```
콘솔보다 logging 라이브러리 사용 추천드립니다

xenex90 replied:
네 logger로 출력하겠습니다.

f-lab-troy commented on everycodeacademy/src/main/java/com/every/everycodeacademy/API/CompileAPI.java:
```diff
@@ -0,0 +1,90 @@
+      e.printStackTrace();
+    }
+
+    // 3. 컴파일 옵션 설정
```
이 부분도 상수 또는 설정 파일로부터 받아오면 좋을 것 같습니다

xenex90 replied:
설정 파일 만들어서 상수 같은 것들 관리해 보겠습니다.

f-lab-troy commented on everycodeacademy/src/main/java/com/every/everycodeacademy/API/CompileAPI.java:
```diff
@@ -0,0 +1,90 @@
+      if (compilationResult == 0) {
+        System.out.println("컴파일이 성공적으로 완료되었습니다.");
+      } else {
+        System.out.println("컴파일 중 오류가 발생하였습니다.");
```
오류 상황에서 예외 발생 시켜서 처리하는걸 고려해보면 좋습니다.

xenex90 replied:
네 수정하겠습니다

f-lab-troy commented on everycodeacademy/src/main/java/com/every/everycodeacademy/API/CompileAPI.java:
```diff
@@ -0,0 +1,90 @@
+    }
+
+    // 삭제 위해 파일 권한 변경
+    javaCompileService.changeClassFilePermissions();
```
중복으로 보입니다

f-lab-troy commented on everycodeacademy/src/main/java/com/every/everycodeacademy/compile/JavaCompileService.java:
```diff
@@ -0,0 +1,107 @@
+
+@Service
+public class JavaCompileService {
+  public boolean deleteJavaFile() {
```
밑의 deleteClassFile 메서드와 같은 로직을 수행하는 것으로 보입니다. 공통 메서드로 분리해보시는 걸 추천드립니다

xenex90 replied:
생각해보고 그렇게 하겠습니다

f-lab-troy commented on everycodeacademy/src/main/java/com/every/everycodeacademy/compile/JavaCompileService.java:
```diff
@@ -0,0 +1,107 @@
+      return false;
+    }
+  }
+
```
위 코멘트와 마찬가지입니다. 중복은 피하시는게 좋습니다.

xenex90 replied:
네 알겠습니다
클래스 파일 삭제 처리, 자바 파일 삭제 파일
메서드는 하나의 기능을 하는게 좋다고 이해해서 이렇게 했었는데
너무 오바해서 적용시킨 케이스 인 것 같습니다

f-lab-troy commented:
전체적으로 고려하시길 추천 드리는 내용입니다. 
- system.out 대신 logback과 같은 로깅 라이브러리 사용
- 문제 발생시 콘솔에 쓰는 대신 예외 발생 시키고 처리
- 테스팅 코드가 직접 작성하신 class와 method를 테스트 하는게 아닌 method안의 로직을 테스트하는 것으로 보입니다. 이런 경우 로직이 바뀌면 테스트도 바뀌어야 하는 상황이 발생하기 때문에 피하시는게 좋습니다. 
- 테스팅 코드안에 system.out은 assert로 변환하여 자동화가 가능하도록 만들어야합니다

xenex90 commented:
네 감사합니다.
말씀해주신 것 참고해서 수정하겠습니다

xenex90 committed:
modify web code compiler

- 스프링 부트 로깅 라이브러리 사용
- system.out -> assert와 로거로 변경
- 에러 발생할 경우 -> 예외 발생 처리
- 테스트 케이스 구조 변경 - 내부 로직 테스트 -> 메서드 테스트

xenex90 committed:
modify web code compiler

- 스프링 부트 로깅 라이브러리 사용
- system.out -> assert와 로거로 변경
- 에러 발생할 경우 -> 예외 발생 처리
- 테스트 케이스 구조 변경 - 내부 로직 테스트 -> 메서드 테스트

xenex90 commented:
변경사항

- 꼭 필요하지 않은 코드 정리
- 스프링 부트 로깅 라이브러리 사용
- system.out -> assert와 로거로 변경
- 에러 발생할 경우 -> 예외 발생 처리
- 테스트 케이스 구조 변경 - 내부 로직 테스트 -> 메서드 테스트

설정파일에 상수 및 중요한 값 저장 하는 것으로 추가/변경도 예정되어 있습니다.

xenex90 committed:
Managed constants

클래스 파일명과 자바 파일명들을 상수 클래스 생성 후 이 클래스에서 관리해서 호출하도록 변경하였습니다.

xenex90 committed:
Merge branch 'main' into everycodeacademy

xenex90 commented on everycodeacademy/src/main/java/com/every/everycodeacademy/compile/JavaCompileService.java:
```diff
@@ -22,10 +23,8 @@ public class JavaCompileService {
-    String filePath = "webCompile.java";
 
-    File file = new File(filePath);
+    File file = new File(Constants.WEB_COMPILE_JAVA_FILE_PATH);
```
상수 클래스 호출로 변경하였습니다.

xenex90 commented on everycodeacademy/src/main/java/com/every/everycodeacademy/compile/JavaCompileService.java:
```diff
@@ -75,52 +74,54 @@ public void deleteClassNJavaFile() {
   }
 
-  public boolean changeJavaFilePermissions() {
+  public boolean changeJavaFilePermissions() throws Exception {
```
try-catch 문이 changeJavaFilePermissions메소드 안에 있긴 하지만 if 조건문으로 정확히 예외 처리 하기 어려워
throws Exception으로 예외 발생 시키고 원 호출 클래스에서 받아서
try-catch 문으로 받아서 예외 발생시키는데 이렇게 하는 것이 맞는 지 궁금합니다.

xenex90 commented on everycodeacademy/src/test/java/com/every/everycodeacademy/API/compileAPITest.java:
```diff
@@ -91,12 +91,18 @@ void getJavaStringToCompileTest() {
+      assertTrue (javaCompileService.changeJavaFilePermissions());
+    } catch (Exception e) {
+      e.printStackTrace();
+      logger.error(e.getMessage());
```
logger.error로 에러시 에러메세지 출력하도록 하였는데
에러 발생한 메소드에서 로깅하는 것이 좋은지
메소드 호출한 클래스나 메소드에서 받은 예외를 처리하면서 그 곳에서 예외처리를 하는 것이 좋은지 잘 모르겠습니다.

xenex90 commented on everycodeacademy/src/main/java/com/every/everycodeacademy/constant/Constants.java:
```diff
@@ -0,0 +1,9 @@
+package com.every.everycodeacademy.constant;
+
+public final class Constants {
```
상수 클래스를 만들어서 webCompile.java 경로명들을 관리하도록 일단 하였습니다.
application.porperties 파일에서 하는 방법도 있다고 하던데 개발, 실서버 등 여러 종류의 서버가 있을 떄 좋다고 해서 상수 클래스로 만들어서 하였습니다.

xenex90 commented:
@f-lab-troy 멘토님

변경하지 않는 자바 클래스 경로명 같은 상수는 상수 클래스 만들어서 호출하도록 변경하였습니다.

f-lab-troy reviewed (approved):
세션때 나누었던 내용(예: 중복, 테스트)은 다른 pr로 수정하시는 것으로 알고 있겠습니다

xenex90 commented:
네 지금 작업 중인데 완료되면 PR 드리도록 하겠습니다.
